### ISSUE TITLE
Dynamic related models for relations

### ISSUE THREAD

**Issue by mvdstam** [laravel/framework#11294]
Hi,

I've come across a situation where it is desirable to dynamically return different models for the same relation. Please consider the following (schematic) example:

```php
<?php

class A extends Model {

  public function doSomethingBasic()
  {
    // Logic specific to A, and abstract enough to apply to all derivates of A
  }
}

class B extends A {

  public function doSomethingBasic()
  {
    // Logic specific to B, which overrides the method of A
  }
}

class C extends A {

  public function doSomethingAdditionalAsWell()
  {
    // Logic specific to C
  }

  public function doSomethingBasic()
  {
    // Logic specific to C, which overrides the method of A
  }
}

class ParentModel extends Model {

  public function children()
  {
    return $this->hasOne(A::class);
  }
}
```

In the example above, all classes are derived from the base `Model` which ships with Eloquent. The `ParentModel` class has a relation with class `A`. Because classes `B` and `C` are derived from `A`, they inherit all properties (including database-level attributes such as the table name and primary key). In my specific situation, the table containing entries for `A` has certain fields which may or may not mean the related model should actually be a child class of `A`, such as application-level logic that `A` doesn't need, but `B` or `C` would.
Based on the data of each row of the table containing `A` entries, the `Collection` (or single model instance) returned from the relation call would be instances of `A`or instances that are derived from `A` (in accordance to the Liskov Substitution).

Basically, this particular problem would be solved if I could simply hook into the part of the routine that creates an instance of the class that is passed into the `Relation` class, and being able to override that logic with my own. I would suggest to alter the relationship methods of Eloquent's `Model` class to make it possible to pass a closure (as an additional method argument) which decides what instance is finally returned. That closure would simply accept a single array of data which would represent a single row of data from the database table.

Thoughts?

**Comment by EvgeniyN** [comment]
Hi, I'm not sure, but (as I think it's SingleTableInheritance) and u can do:
class A extends Model
{
 ...
    public function newFromBuilder($attributes = array())
    {
        switch( $id = array_get($attributes, 'id') ){
            case 75:
                return new \B($attributes);

            case 368:
                return new \C($attributes);

            default:
                return parent::newFromBuilder($attributes);
        }
    }
...
Sry for cucumber eng ))

**Comment by GrahamCampbell** [comment]
I'm sorry, but we don't process proposal issues. Feel free to discuss on slack and/or the forums, and possibly send a PR depending on the results of those discussions.
